Implement Resource Protection in General Resources

carlaKC opened the pull request:
Opening this up for discussion - not final design.

Possible approach for #45

elnosh commented on ln-resource-mgr/src/forward_manager.rs:
```diff
@@ -337,6 +340,7 @@ impl ReputationManager for ForwardManager {
                 v.insert(TrackedChannel {
                     capacity_msat,
                     incoming_direction: IncomingChannel::new(
+                        channel_id,
```
wondering if we should also assign slots for every new added channel here in `add_channel`? iiuc we'll assign them at the time we are adding an htlc if the channel is not found in the `RestrictedBucket.candidate_slots` map.

carlaKC replied:
> may_add_htlc won't need mutable references since they wouldn't be assigning the slots at the moment

Good point, will update 👍

elnosh commented on ln-resource-mgr/src/lib.rs:
```diff
@@ -278,7 +284,11 @@ impl AllocationCheck {
                     .resources_available(htlc_amt_msat)
                 {
-                    Ok(ResourceBucketType::General)
+                    if self.general_eligible {
+                        Ok(ResourceBucketType::General)
+                    } else {
+                        Err(FailureReason::GeneralResourceLimit)
+                    }
```
just to make sure this is the intended behavior, if the channel has exhausted its general slots, they don't get a shot at the congestion bucket?

carlaKC replied:
Discussed offline - we do indeed want to reject a pair that's taken up all of its slots even when general still has some space, this is how we get some DOS protections against people trying to spam up general.

elnosh commented on ln-resource-mgr/src/incoming_channel.rs:
```diff
@@ -20,19 +23,27 @@ pub(super) struct IncomingChannel {
     /// accountable from peers that have sufficient reputation.
     pub(super) protected_bucket: BucketParameters,
+
+    pub(super) general_resources: RestrictedBucket,
```
not sure what to suggest here but it feels weird to have a `general_bucket` and a `general_resources` field here in the incoming channel.

carlaKC commented on ln-resource-mgr/src/forward_manager.rs:
```diff
@@ -401,6 +405,7 @@ impl ReputationManager for ForwardManager {
 
         let fwd_outcome = match fwd_outcome {
             Ok(fwd_sucess) => {
+                // TODO: we need to add to the incoming channel here but it's not avail
```
> Also since we need to surface the AllocationCheck to report that htlc forward, I don't see a way around it bc we'd still need to do them separately (get the AllocationCheck to check the outcome and then add it)

Yeah indeed. Was wondering if we want to think of a different design that doesn't break things up into get check / then assign but I don't see an obvious way of avoiding it actually.

carlaKC commented on ln-resource-mgr/src/lib.rs:
```diff
@@ -263,6 +267,8 @@ impl AllocationCheck {
                             .general_bucket
                             .resources_available(htlc_amt_msat)
                         {
+                            // TODO: should we limit high reputation peers to their allocated
+                            // general resources?
```
Meaning, should we check `general_eligible` here. 

But on second look I don't think so - once somebody's got reputation we should basically just let them use whatever IMO.

elnosh commented:
all those allocations along with this which you had already mentioned before :man_facepalming: makes sense to go with opportunistic approach (:  

> it's unlikely that we need to use all of these allocations.

my bad I suggested you go down this path

carlaKC commented:
> all those allocations along with this which you had already mentioned before

Only gets real when you spend the time to realize it's big bad N-squared 🐺 

Always time well spent taking a look at alternatives, and we'll have a much better documented approach for having done it!

carlaKC commented:
Think I've addressed most of the review comments. Major changes since last review are:
* Stricter bucket assignments: eg, still require `general_eligible` if you have reputation because you may be the one attacking us
* Table driven test: added to try enumerate all `bucket_outcome` possibilities, left this in its own commit because it's big but can squash it if we want

elnosh commented on ln-resource-mgr/src/lib.rs:
```diff
@@ -764,73 +736,271 @@ mod tests {
+                case
+            },
+            {
+                let mut case = TestCase::new(
+                    "unaccountable_no_upgrade_congestion_available",
+                    AccountableSignal::Unaccountable,
+                    Ok(ResourceBucketType::Congestion),
+                );
+                case.upgradable = false;
+                case.set_congestion_available();
+                case
+            },
```
not strictly related to this PR but looking at this test case it made me realize that this should instead be failing? Getting a bucket outcome of congestion will cause the HTLC to be upgraded to accountable but if it the htlc is not upgradable, it should fail.

elnosh reviewed (commented):
mostly LGTM

carlaKC commented:
* Added requirement for `upgradable` for congestion use: [diff](https://github.com/carlaKC/jam-ln/compare/10e69b2a8913d72773060e2d564487ddb1185aff..e289cba0c3b995c1bf6c7b12854a1687278f966f)
* Rebased to clear `Cargo.lock` conflict: [diff](https://github.com/carlaKC/jam-ln/compare/e289cba0c3b995c1bf6c7b12854a1687278f966f..951e7fd2f9e1ef708457d232ab2b10deb6197752)
